PSReadLine/DynamicHelp.cs

```diff
@@ -0,0 +1,288 @@
+                        console.BlankRestOfLine();
+                    }
+
+                    // Explicit newline so consoles see each row as distinct lines, but skip the
+                    // last line so we don't scroll.
+                    if (row != (this.Rows - 1) || !menuSelect)
+                    {
+                        AdjustForPossibleScroll(1);
+                        MoveCursorDown(1);
+                    }
+                }
+
+                bool extraPreRowsCleared = false;
+                if (previewBlock != null)
+                {
+                    if (Rows < previewBlock.Rows)
+                    {
+                        // If the last menu row took the whole buffer width, then the cursor could be pushed to the
+                        // beginning of the next line in the legacy console host (NOT in modern terminals such as
+                        // Windows Terminal, VSCode Terminal, or virtual-terminal-enabled console host). In such a
+                        // case, there is no need to move the cursor to the next line.
+                        //
+                        // If that is not the case, namely 'CursorLeft != 0', then the rest of the last menu row was
+                        // erased, but the cursor was not moved to the next line, so we will move the cursor.
+                        if (console.CursorLeft != 0)
+                        {
+                            // There are lines from the previous rendering that need to be cleared,
+                            // so we are sure there is no need to scroll.
+                            MoveCursorDown(1);
+                        }
+
+                        Singleton.WriteBlankLines(previewBlock.Rows - Rows);
+                        extraPreRowsCleared = true;
+                    }
+                }
+
+                // if the menu has moved, we need to clear the lines under it
+                if (bufferEndPoint.Y < PreviousTop)
+                {
+                    // In either of the following two cases, we will need to move the cursor to the next line:
+                    //  - if extra rows from previous menu were cleared, then we know the current line was erased
+                    //    but the cursor was not moved to the next line.
+                    //  - if 'CursorLeft != 0', then the rest of the last menu row was erased, but the cursor
+                    //    was not moved to the next line.
+                    if (extraPreRowsCleared || console.CursorLeft != 0)
+                    {
+                        // There are lines from the previous rendering that need to be cleared,
+                        // so we are sure there is no need to scroll.
+                        MoveCursorDown(1);
+                    }
+
+                    Singleton.WriteBlankLines(PreviousTop - bufferEndPoint.Y);
+                }
```

Comment:
Same here, for `Menu`, the `Top` could change given a different selection, but that won't happen for this type.

Reply:
refactored